### PR TITLE
MGMT-5422: AgentClusterInstall - ControllerLogs conditions

### PR DESCRIPTION
**Pull request by RazRegev**
Two conditions were added - "LogCollectionCompleted" and
"LogCollectionStopped", contain the same data bug exists
for user convenience.
When collection is completed, it will expose the download
url in the debugInfo under logsURL field.

### CODE REVIEW

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 1]
```diff
@@ -1073,3 +1076,45 @@ func (r *ClusterDeploymentsReconciler) ensureOwnerRef(ctx context.Context, cd *h
...
+		reason = LogCollectionCollectingReason
+		message = LogCollectionCollectingMsg
+	default:
+		status = corev1.ConditionFalse
+		reason = LogCollectionUnavailableReason
+		message = LogCollectionUnavailableMsg
+	}
+
+	setClusterCondition(&clusterInstall.Status.Conditions, hivev1.ClusterInstallCondition{
+		Type:    LogCollectionCompletedCondition,
```
Comment (rollandf) [inline internal/controller/controllers/clusterdeployments_controller.go:1109]: Not sure I understand why we need two conditions

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 2]
```diff
@@ -1073,3 +1076,45 @@ func (r *ClusterDeploymentsReconciler) ensureOwnerRef(ctx context.Context, cd *h
...
 }
+
+func (r *ClusterDeploymentsReconciler) clusterControllerLogs(clusterInstall *hiveext.AgentClusterInstall, cluster *common.Cluster) {
+	var status corev1.ConditionStatus
+	var reason, message string
+	switch {
+	case !swag.IsZero(cluster.ControllerLogsCollectedAt):
+		status = corev1.ConditionTrue
+		reason = LogCollectionCompletedReason
+		downloadURL := fmt.Sprintf("%s%s/clusters/%s/logs",
```
Comment (rollandf) [inline internal/controller/controllers/clusterdeployments_controller.go:1087]: This will not work with local auth

**Review by rollandf — CHANGES_REQUESTED** [review]
- Please also update the documentation
- Also add this condition in `setClusterConditionsUnknown`

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 1 above)
Comment (RazRegev) [reply internal/controller/controllers/clusterdeployments_controller.go:1120]: One for stop, second for completed - for two sections in hive code (we already talked about that, what am I missing?)

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 2 above)
Comment (RazRegev) [reply internal/controller/controllers/clusterdeployments_controller.go:1094]: fixed

File: cmd/main.go [hunk 3]
```diff
@@ -461,6 +461,8 @@ func main() {
 				HostApi:          hostApi,
 				CRDEventsHandler: crdEventsHandler,
 				Manifests:        manifestsApi,
+				ServiceBaseURL:   Options.BMConfig.ServiceBaseURL,
+				AuthHandler:      authHandler,
```
Comment (rollandf) [inline cmd/main.go:465]: No need for the whole `AuthHandler`, AuthType should be enough

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 4]
```diff
@@ -1073,3 +1082,67 @@ func (r *ClusterDeploymentsReconciler) ensureOwnerRef(ctx context.Context, cd *h
...
+	var reason, message string
+	switch {
+	case !swag.IsZero(cluster.ControllerLogsCollectedAt):
+		status = corev1.ConditionTrue
+		reason = LogCollectionCompletedReason
+		downloadURL, err := r.getControllerLogsDownloadURL(cluster)
+		if err != nil {
+			return err
+		}
+		message = fmt.Sprintf("%s - download URL: %s", LogCollectionCompletedMsg, downloadURL)
```
Comment (rollandf) [inline internal/controller/controllers/clusterdeployments_controller.go:1100]: The URL should not be in the message but in a new field under AgentClusterInstall Status.
debugInfo:
    logsURL: http://10.1.37.42:6000/api/assisted-install/v1/.........

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 1 above)
Comment (rollandf) [reply internal/controller/controllers/clusterdeployments_controller.go:1120]: Hive is not looking for Logs conditions. As per Avishai request we can leave it with two conditions to make it clear.

**Review by rollandf — CHANGES_REQUESTED** [review]
Nice!

- Need to set to Unknown in `setClusterConditionsUnknown`
- Update documentation in `docs/kube-api-conditions.md`
- The URL should in be a field in a new struct `DebugInfo` under Status. See #1776 for struct definition

File: docs/kube-api-conditions.md [hunk 5]
```diff
@@ -32,6 +32,19 @@ AgentClusterInstall supported condition types are: `SpecSynced`, `RequirementsMe
 |Stopped|True|InstallationCancelled|The installation has stopped because it was cancelled|if the cluster status is "cancelled"|
 |Stopped|True|InstallationCompleted|The installation has stopped because it completed successfully|if the cluster status is "installed"|
 |Stopped|False|InstallationNotStopped|The installation is waiting to start or in progress|If the cluster status is not "error", "cancelled" or "installed|
+||||||
+|LogCollectionCompleted|True|Completed|Logs are available|after installation completed or failed|
```
Comment (rollandf) [inline docs/kube-api-conditions.md:36]: "after installation completed or failed and logs are collected"

File: docs/kube-api-conditions.md [hunk 6]
```diff
@@ -32,6 +32,19 @@ AgentClusterInstall supported condition types are: `SpecSynced`, `RequirementsMe
...
+|LogCollectionCompleted|False|Requested|Logs have been requested but not yet collected|a request for logs has been made before gathering|
+|LogCollectionCompleted|False|Collecting|Logs are currently being collected|after installation completed or failed|
+|LogCollectionCompleted|False|Timeout|Log collection has failed due to timeout|after installation has completed or failed|
+|LogCollectionCompleted|False|Unavailable|Log collection not yet started|before and during installation|
+||||||
+|LogCollectionStopped|True|Completed|Logs are available|after installation completed or failed|
```
Comment (rollandf) [inline docs/kube-api-conditions.md:42]: "after installation completed or failed and logs are collected"

**Review by rollandf — COMMENTED** [review]
Looks good! A few doc comments, and rebase when other fixes are merged

**Comment by filanov** [comment]
Overall looks good, did you check if it works and can you put an example of the CR with the logs?

**Comment by RazRegev** [comment]
> Overall looks good, did you check if it works and can you put an example of the CR with the logs?

yes sure, in a few moment I will upload an example

**Comment by filanov** [comment]
I think that you are missing a cleanup of the logs if the installation is successful https://github.com/openshift/assisted-service/blob/master/internal/controller/controllers/clusterdeployments_controller.go#L148

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 7]
```diff
@@ -164,8 +164,8 @@ func (r *ClusterDeploymentsReconciler) Reconcile(ctx context.Context, req ctrl.R
 		if !r.isSNO(clusterInstall) {
 			return r.createNewDay2Cluster(ctx, req.NamespacedName, clusterDeployment, clusterInstall)
 		}
-		// cluster is installed and SNO. Clear EventsURL
-		clusterInstall.Status.DebugInfo.EventsURL = ""
+		// cluster  is installed and SNO.
```
Comment (filanov) [inline internal/controller/controllers/clusterdeployments_controller.go:167]:
```suggestion
		// cluster is installed and SNO.
```

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 8]
```diff
@@ -1053,6 +1063,18 @@ func (r *ClusterDeploymentsReconciler) populateEventsURL(log logrus.FieldLogger,
 	return nil
 }
 
+func (r *ClusterDeploymentsReconciler) populateLogsURL(log logrus.FieldLogger, clusterInstall *hiveext.AgentClusterInstall, c *common.Cluster) error {
+	if *c.Status != models.ClusterStatusInstalled && !swag.IsZero(c.ControllerLogsCollectedAt) {
```
Comment (rollandf) [inline internal/controller/controllers/clusterdeployments_controller.go:1067]: No need for the check for `!swag.IsZero(c.ControllerLogsCollectedAt)` , it is done in `setControllerLogsDownloadURL`

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 9]
```diff
@@ -180,6 +180,10 @@ func (r *ClusterDeploymentsReconciler) Reconcile(origCtx context.Context, req ct
 			return r.createNewDay2Cluster(ctx, log, req.NamespacedName, clusterDeployment, clusterInstall)
 		}
 		// cluster is installed and SNO nothing to do.
+		if updateErr := r.Status().Update(ctx, clusterInstall); updateErr != nil {
```
Comment (rollandf) [inline internal/controller/controllers/clusterdeployments_controller.go:183]: This is not needed anymore here. Rebase issue

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 8 above)
Comment (RazRegev) [reply internal/controller/controllers/clusterdeployments_controller.go:1067]: it currently isn't done there but I will add it to there

**Comment by rollandf** [comment]
Looks good! let see what is going on with CI

**Review by filanov — APPROVED** [review]

**Review by filanov — APPROVED** [review]

**Comment by RazRegev** [comment]
new PR will be opened without the conditions.

**Pull request closed without merging**